vllm/lora/punica.py

```diff
@@ -1,207 +1,374 @@
-# Based on code from https://github.com/punica-ai/punica
+"""
+Based on:
+Chen, L., Ye, Z., Wu, Y., Zhuo, D., Ceze, L., & Krishnamurthy, A. (2023). 
+Punica: Multi-Tenant LoRA Serving. 
+https://arxiv.org/abs/2310.18547
+"""
 
-from typing import Optional
+from typing import Optional, Tuple
 
 import torch
 
-from vllm import _custom_ops as ops
-from vllm.platforms import current_platform
+from vllm.lora.ops.bgmv_expand import bgmv_expand
+from vllm.lora.ops.bgmv_expand_slice import bgmv_expand_slice
+from vllm.lora.ops.bgmv_shrink import bgmv_shrink
+from vllm.lora.ops.sgmv_expand import sgmv_expand
+from vllm.lora.ops.sgmv_expand_slice import sgmv_expand_slice
+from vllm.lora.ops.sgmv_shrink import sgmv_shrink
 
 
-def _check_punica_support():
-    if ops.is_custom_op_supported("_punica_C::dispatch_bgmv"):
-        return
+def _compute_meta(
+    token_lora_tensor: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, int, int, ]:
+    """
+    Get the information required for the sgmv kernel. With the  features:
+    1. If consecutive requests in the batch use the same LoRA, this function 
+    will combine them into a single request, improving sgmv kernel inference 
+    performance.
+    2. At the beginning of each prefill stage inference, recalculations are 
+    needed based on the input, but only once. 
+    """
 
-    if current_platform.get_device_capability() < (8, 0):
-        raise ImportError(
-            "punica LoRA kernels require compute capability >= 8.0")
-    else:
-        raise ImportError(
-            "punica LoRA kernels could not be imported. If you built vLLM "
-            "from source, make sure VLLM_INSTALL_PUNICA_KERNELS=1 env var "
-            "was set.")
+    lora_indices_tensor, seq_length_tensor = torch.unique_consecutive(
+        token_lora_tensor, return_counts=True)
+    cum_result = torch.cumsum(seq_length_tensor, dim=0)
+    b_seq_start_tensor = torch.zeros_like(seq_length_tensor)
+    b_seq_start_tensor[1:].copy_(cum_result[:-1])
+    max_length = seq_length_tensor.max().item()
+    batch_size = lora_indices_tensor.size(0)
+    return (
+        b_seq_start_tensor,
+        seq_length_tensor,
+        lora_indices_tensor,
+        batch_size,
+        max_length,
+    )
+
+
+class PrefillHelper:
+    """PrefillHelper is designed to manage and provide metadata for the sgmv 
+    kernel during  prefill stage, utilizing the singleton pattern to guarantee 
+    the existence of only one instance of the class.
+    """
+    _instance: Optional["PrefillHelper"] = None
+    initialized: bool
+
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+            cls._instance.initialized = False
+        return cls._instance
```

Comment:
I don't like this pattern. It will still lead to a global reference being left in memory. Can we just pass the prefill helper to the methods - or even better, just make a normal class like `PunicaWrapper` or something that will hold the state and have `shrink_prefill`, `shrink_decode` etc. as methods.